src/operator/image/image_random.cu

```diff
@@ -99,18 +111,242 @@ void ToTensorImplCUDA(mshadow::Stream<gpu> *s,
+                KERNEL_ASSIGN(output[c][h][w], req,
+                              (input[c][h][w] - mean) / std);
+            }
+        }
+    }
+}
+
+// Normalize Kernel for 4D input
+template<typename xpu, typename DType>
+__global__ void
+__launch_bounds__(cuda::kMaxThreadsPerBlock, 1)
+NormalizeCudaKernel(const Tensor<xpu, 4, DType> input,
+                    const Tensor<xpu, 4, DType> output,
+                    const int req,
+                    const int N,
+                    const int H,
+                    const int W,
+                    const int C,
+                    const float mean_d0,
+                    const float mean_d1,
+                    const float mean_d2,
+                    const float std_d0,
+                    const float std_d1,
+                    const float std_d2) {
+    // We process one image per thread block.
+    const int n = blockIdx.x;
+
+    float mean = mean_d0;
+    float std = std_d0;
+    for (int c = 0; c < C; ++c) {
+        switch (c) {
+            case 0 : mean = mean_d0;
+                     std = std_d0;
+                     break;
+            case 1 : mean = mean_d1;
+                     std = std_d1;
+                     break;
+            case 2 : mean = mean_d2;
+                     std = std_d2;
+                     break;
+        }
+        for (int h = threadIdx.y; h < H; h += blockDim.y) {
+            for (int w = threadIdx.x; w < W; w += blockDim.x) {
+                KERNEL_ASSIGN(output[n][c][h][w], req,
+                              (input[n][c][h][w] -  mean) / std);
+            }
+        }
+    }
+}
+
+template<typename DType, typename T>
+void NormalizeImplCUDA(mshadow::Stream<gpu> *s,
+                       const T input,
+                       const T output,
+                       const int req,
+                       const float mean_d0,
+                       const float mean_d1,
+                       const float mean_d2,
+                       const float std_d0,
+                       const float std_d1,
+                       const float std_d2) {
+    int blocks, H, W, C, N;
+    cudaStream_t stream = mshadow::Stream<gpu>::GetStream(s);
+    if (std::is_same<T, Tensor<gpu, 3, DType>>::value) {
+        // 3D Input - (C, H, W)
+        N = 0;
+        C = input.size(0);
+        H = input.size(1);
+        W = input.size(2);
+        blocks = 1;
+    } else {
+        // 4D Input - (N, C, H, W)
+        N = input.size(0);
+        C = input.size(1);
+        H = input.size(2);
+        W = input.size(3);
+        blocks = N > 0 ? N : 1;
+    }
+    // One block per image.
+    NormalizeCudaKernel<gpu, DType>
+        <<<blocks, dim3(H, cuda::kMaxThreadsPerBlock / H), 0, stream>>>(input, output,
+            req, N, H, W, C, mean_d0, mean_d1, mean_d2,
+            std_d0, std_d1, std_d2);
+    MSHADOW_CUDA_POST_KERNEL_CHECK(NormalizeCudaKernel);
+}
+
+// Normalize Backward Kernel for 3D input
+template<typename xpu, typename DType>
```

Comment:
Still think 1D kernel is more efficient

```c
// launch with 1D, NormalizeBackwardCudaKernel<gpu, DType>
//        <<<blocks, dim3(cuda::kMaxThreadsPerBlock, 1), 0, stream>>>(out_grad, in_grad,
//            req, N, H, W, C, std_d0, std_d1, std_d2);
template<typename xpu, typename DType>
__global__ void
__launch_bounds__(cuda::kMaxThreadsPerBlock, 1)
NormalizeBackwardCudaKernel(const Tensor<xpu, 3, DType> out_grad,
                            const Tensor<xpu, 3, DType> in_grad,
                            const int req,
                            const int N,
                            const int H,
                            const int W,
                            const int C,
                            const float std_d0,
                            const float std_d1,
                            const float std_d2) {
    for (int i = threadIdx.x; i < H * W * C; ++threadIdx.x) {
        float std = std_d0;
        int c = i % 3;
        switch (c) {
            case 0 : break;
            case 1 : std = std_d1;
                     break;
            case 2 : std = std_d2;
                     break;
        KERNEL_ASSIGN(*(in_grad.data_ + i), req,
                              *(out_grad.data_ + i) * (1 / std));
        }
    }
}
```

Reply:
Fixed and made it 1D kernel. Thanks for the pointers.